text/0000-cortex-m-targets.md

```diff
@@ -0,0 +1,595 @@
+  Tag_ABI_PCS_RO_data: PC-relative
+  Tag_ABI_PCS_GOT_use: GOT-indirect
+  Tag_ABI_FP_denormal: Needed
+  Tag_ABI_FP_exceptions: Needed
+  Tag_ABI_FP_number_model: IEEE 754
+  Tag_ABI_align_needed: 8-byte
+  Tag_ABI_align_preserved: 8-byte, except leaf SP
+  Tag_ABI_optimization_goals: Prefer Debug
+  Tag_CPU_unaligned_access: v6
+  Tag_ABI_FP_16bit_format: IEEE 754
+```
+
+### Building a binary
+
+Two extra steps are required to build a binary:
+
+- Tell Cargo which linker to use.
+- Supply a linker script to the linker.
+  
+[vector table]: http://infocenter.arm.com/help/topic/com.arm.doc.dui0552a/BABIFJFG.html
+
+```
+$ rustup target add cortex-m3
+
+$ cargo new --bin app && cd $_
+
+# Tell Cargo about the linker and the linker script.
+$ mkdir .cargo && edit .cargo/config && cat .cargo/config
+```
+
+``` toml
+[build]
+rustflags = ["-C", "link-args=-Tlayout.ld"]
+
+[target.cortex-m3]
+linker = "arm-none-eabi-gcc"
+```
+
+```
+# The is the linker script
+$ edit layout.ld && cat layout.ld
+/* Device-specific memory regions */
+MEMORY
+{
+  FLASH : ORIGIN = 0x08000000, LENGTH = 128K
+  SRAM  : ORIGIN = 0x20000000, LENGTH = 8K
+}
+
+/* Omitted:
+   - Definition and placement of the vector table and other program sections
+   - Addresses of memory mapped peripherals */
+
+$ edit src/main.rs && cat src/main.rs
+```
+
+``` rust
+#![no_main]
+#![no_std]
+
+// Program entry point
+fn start() -> ! {
+    // ...
+}
+
+// Omitted: definition of lang items, placing `start` in the vector table, etc.
+```
+
+```
+$ cargo build --target cortex-m3
+
+$ file target/cortex-m3/debug/app
+target/cortex-m3/debug/app ELF 32-bit LSB executable, ARM, EABI5 version 1 (SYSV), statically linked, not stripped
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None so far.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Don't do this
+
+Leave it up to the user figure out how to set up their development environment.
+
+Luckily, the author is currently writing [extensive documentation] on the area of Rust on
+microcontrollers, and also has developed tools that make cross compiling [freestanding crates] and
+[compiler-rt] relatively easy. However, adding these targets to the compiler will make the setup
+process simpler (less steps, requires less tooling, etc).
+
+[extensive documentation]: http://japaric.github.io/copper
+[freestanding crates]: https://github.com/japaric/xargo
+[compiler-rt]: https://github.com/japaric/compiler-rt.rs
+
+## Use the hard-float CC instead of the soft-float one
+
+The main proposal opted for the soft-float CC, favoring interoperability over performance. However,
+using the hard-float CC to favor performance is also a valid alternative.
+
+Yet another alternative is to provide both CC variants for the targets that support them. For
+example, instead of a single `cortex-m4f` target now there would be a `cortex-m4f` target for the
+soft-float CC and a `cortex-m4f-hf` target for the hard-float CC.
+
+## Less targets
```

Comment:
I'm assuming your comment is referring to the "Use the hard-float CC instead of the soft-float one" alternative and not the "Less targets" alternative.  If so, I agree.  I dont' see any reason why all variants cannot be specified and require the user disambiguously specify which ABI they wish to compiler for.